File: MRacDemo/MRacDemo/MMenu/MMenuViewController.m
//
//  MMenuViewController.m
//  MRacDemo
//
//  Created by mal on 2019/9/28.
//  Copyright © 2019 mal. All rights reserved.
//

#import "MMenuViewController.h"

#define MCellIdentifier @"MCellIdentifier"

@interface MMenuViewController ()

@property (nonatomic, strong) NSMutableArray *dataSource;

@end

@implementation MMenuViewController

//MARK: - get & set
- (NSMutableArray *)dataSource
{
    if (_dataSource == nil)
    {
        _dataSource = [[NSMutableArray alloc] init];
    }
    return _dataSource;
}

//- (UITableView *)tableView
//{
//    if (_tableView) {
//        <#statements#>
//    }
//}

//MARK: - public method
- (void)addItemWithClass:(Class)aClass title:(NSString *)title turnType:(MTurnType)type
{
    [self.dataSource addObject:[MMenuItem itemWithClass:aClass title:title turnType:type]];
}

- (void)addItemWithClass:(Class)aClass title:(NSString *)title
{
    [self addItemWithClass:aClass title:title turnType:M_Push];
}

- (void)addItems:(NSArray<MMenuItem *> *)items
{
    [self.dataSource addObjectsFromArray:items];
}

//MARK: - system method
- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setUpView];
}

//MARK: - UI
- (void)setUpView
{
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:MCellIdentifier];
}

//MARK: - UITableViewDataSource & UITableViewDelegate
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:MCellIdentifier];
    cell.textLabel.text = [self.dataSource[indexPath.row] itemTitle];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    MMenuItem *item = self.dataSource[indexPath.row];
    id obj = [[item.itemClass alloc] init];
    if ([obj isKindOfClass:[UIViewController class]])
    {
        UIViewController *vc = (UIViewController *)obj;
        vc.view.backgroundColor = [UIColor whiteColor];
        vc.title = item.itemTitle;
        if (item.turnType == M_Push)
        {
            if (self.navigationController)
            {
                [self.navigationController pushViewController:vc animated:YES];
            }
        }
        else if (item.turnType == M_Present)
        {
            vc.modalPresentationStyle = UIModalPresentationFullScreen;
            [self presentViewController:vc animated:YES completion:nil];
        }
    }
}

@end

File: MAlgorithmDemo/MAlgorithmDemo/MMenuViewController.swift
//
//  MMenuViewController.swift
//  MAlgorithmDemo
//
//  Created by mal on 2019/10/17.
//  Copyright © 2019 mal. All rights reserved.
//

import UIKit

let KCellIdentifier = "cellIdentifier"


class MMenuItem {
    
    var itemTitle:String = ""
    var itemClass:UIViewController.Type?
    class func ItemWith(title:String, itemClass:UIViewController.Type) -> MMenuItem{
        
        let item = MMenuItem()
        item.itemTitle = title
        item.itemClass = itemClass
        return item
    }
}

class MMenuViewController: UITableViewController {

    var dataSource:[MMenuItem] = [MMenuItem]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: KCellIdentifier)
    }

    // MARK: - Table view data source
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
       
        return self.dataSource.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: KCellIdentifier)!
        let item = self.dataSource[indexPath.row]
        cell.textLabel?.text = item.itemTitle
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let item = self.dataSource[indexPath.row]
        
        if let vc = item.itemClass?.init(){

            if let navi = self.navigationController{

                vc.view.backgroundColor = UIColor.white
                navi.pushViewController(vc, animated: true)
            }
        }
    }
    
    func mAddItem(item:MMenuItem){
        
        self.dataSource.append(item)
    }
}
